src/pybind/mgr/mirroring/fs/dir_map/policy.py

```diff
@@ -0,0 +1,363 @@
+                                        (len(self.instance_to_dir_map) - nr_dead_instances))
+                if dirs_per_instance == 0:
+                    dirs_per_instance += 1
+                shuffle_dirs = []
+                # super set of directories which are candidates for shuffling -- choose
+                # those which can be shuffle rightaway (others will be shuffled when
+                # they reach idle state).
+                shuffle_dirs_ss = self.shuffle(dirs_per_instance, include_stalled_dirs)
+                if include_stalled_dirs:
+                    return shuffle_dirs_ss
+                for dir_path in shuffle_dirs_ss:
+                    dir_state = self.dir_states[dir_path]
+                    if self.set_state(dir_state, State.STATE_SHUFFLING):
+                        shuffle_dirs.append(dir_path)
+                log.debug(f'remapping directories: {shuffle_dirs}')
+                return shuffle_dirs
+
+    def remove_instances(self, instance_ids):
+        with self.lock:
+            return self._remove_instances(instance_ids)
+
+    def _remove_instances(self, instance_ids):
+        log.debug(f'removing instances: {instance_ids}')
+        shuffle_dirs = []
+        for instance_id in instance_ids:
+            if not instance_id in self.instance_to_dir_map:
+                continue
+            if not self.instance_to_dir_map[instance_id]:
+                self.instance_to_dir_map.pop(instance_id)
+                continue
+            self.dead_instances.append(instance_id)
+            dir_paths = self.instance_to_dir_map[instance_id]
+            log.debug(f'force shuffling instance_id {instance_id}, directories {dir_paths}')
+            for dir_path in dir_paths:
+                dir_state = self.dir_states[dir_path]
+                if self.is_state_scheduled(dir_state, State.STATE_DISASSOCIATING):
+                    log.debug(f'dir_path {dir_path} is disassociating, ignoring...')
+                    continue
+                log.debug(f'shuffling dir_path {dir_path}')
+                if self.set_state(dir_state, State.STATE_SHUFFLING, True):
+                    shuffle_dirs.append(dir_path)
+        log.debug(f'shuffling {shuffle_dirs}')
+        return shuffle_dirs
+
+    def dir_status(self, dir_path):
+        with self.lock:
+            dir_state = self.dir_states.get(dir_path, None)
+            if not dir_state:
+                raise MirrorException(-errno.ENOENT, f'{dir_path} is not tracked')
+            res = {} # type: Dict
+            if dir_state.stalled:
+                res['state'] = 'stalled'
+                res['reason'] = 'no mirror daemons running'
+            elif dir_state.state == State.STATE_ASSOCIATING:
+                res['state'] = 'mapping'
+            else:
+                state = None
+                dstate = dir_state.state
+                if dstate == State.STATE_ASSOCIATING:
+                    state = 'mapping'
```

Comment:
It's handled in init() and add_directory(). The rationale behind that being if a directory is already mapped to an instance, the state is cut-shorted to ASSOCIATING rather than INITIALIZING (which assigned a directory to an instance).